File: javascript/list-ops/list-ops.js
// Implement basic list operations.

// append -> given two lists, add all items in the second list to the end of the first list
// concatenate -> given a series of lists, combine all items in all lists into one flattened list
// filter -> given a predicate and a list, return the list of all items for which predicate(item) is True
// length -> given a list, return the total number of items within it
// map -> given a function and a list, return the list of the results of applying function(item) on all items
// oldl -> given a function, a list, and initial accumulator, fold (reduce) each item into the accumulator from the left using function(accumulator, item)
// foldr -> given a function, a list, and an initial accumulator, fold(reduce) each item into the accumulator from the right using function (item, accumulator)
// reverse -> given a list, return a list with all the original items, but in reversed order

export class List {
  constructor(elements = []) {
    this.values = elements;
  }

  append(otherList) {
    // Step 1: Create a new array that contains all elements from both lists.
    const combinedValues = [...this.values, ...otherList.values];

    // Step 2: Return a new List that uses the combined array.
    return new List(combinedValues);
  }

  concat(listOfLists) {
    // Step 1: Start with a new List that contains the same elements as the current List.
    let newList = new List([...this.values]);

    // Step 2: Iterate over each List in the listOfLists.
    for (let list of listOfLists.values) {
      // Step 3: Check if the current element is a List.
      if (list instanceof List) {
        // If it is, recursively append its elements to newList.
        newList = newList.concat(list);
      } else {
        // If it's not, append it directly to newList.
        newList.values = [...newList.values, list];
      }
    }

    // Step 4: Return the final List.
    return newList;
  }

  filter(predicate) {
    // Step 1: Initialize an empty array to store the values for the new list.
    let filteredListValues = [];

    // Step 2: Iterate over each value in the original list.
    for (let i = 0; i < this.values.length; i++) {
      // Step 3: Apply the predicate to the current value.
      // If the predicate returns true, add the current value to the new list.
      if (predicate(this.values[i])) {
        filteredListValues = [...filteredListValues, this.values[i]];
      }
    }

    // Step 4: Create and return a new List that uses the filtered values.
    return new List(filteredListValues);
  }

  map(func) {
    // Step 1: Initialize an empty array to hold the result.
    const result = [];

    // Step 2: Use a for loop to iterate over `this.values`.
    for (let i = 0; i < this.values.length; i++) {
      // Step 3: Apply the function `func` to the current element,
      // and append the result to the `result` array.
      result.push(func(this.values[i]));
    }

    // Step 4: Return a new `List` instance with the result.
    return new List(result);
  }

  length() {
    // Step 1: Initialize a counter variable to 0
    let length = 0;

    // Step 2: Iterate over each value in the list.
    for (let i = 0; i < this.values.length; i++) {
      // Step 3: For each value, increment the counter by 1.
      length += 1;
    }

    // Step 4: Return the counter, which represents the length of the list.
    return length;
  }

  foldl(func, initial) {
    // Step 1: Initialize the accumulator to the initial value
    let accumulator = initial;

    // Step 2: Use a for loop to iterate over `this.values`
    for (let i = 0; i < this.values.length; i++) {
      // Step 3: Apply the function `func` to the accumulator and the current element,
      // and assign the result to the accumulator
      accumulator = func(accumulator, this.values[i]);
    }

    // Step 4: Return the accumulator
    return accumulator;
  }

  foldr(func, initial) {
    // Step 1: Initialize the accumulator to the initial value
    let accumulator = initial;

    // Step 2: Use a for loop to iterate over `this.values` in reverse order
    for (let i = this.values.length - 1; i >= 0; i--) {
      // Step 3: Apply the function `func` to the accumulator and the current element,
      // and assign the result to the accumulator
      accumulator = func(accumulator, this.values[i]);
    }

    // Step 4: Return the accumulator
    return accumulator;
  }

  reverse() {
    // Step 1: Initialize an empty array for the result
    const result = [];

    // Step 2: Use a for loop to iterate over `this.values` in reverse order
    for (let i = this.values.length - 1; i >= 0; i--) {
      // Step 3: Add the current element to the `result` array
      result.push(this.values[i]);
    }

    // Step 4: Return a new `List` instance with the result
    return new List(result);
  }
}

File: python/list-ops/list_ops.py
def append(list1, list2):
    """
    Appends two lists using list concatenation.

    This function creates a new list by concatenating list1 and list2.
    It does not modify the original lists.

    Args:
    - list1 (list): The first list to be concatenated.
    - list2 (list): The second list to be concatenated.

    Returns:
    - list: A new list containing all elements of list1 followed by all elements of list2.
    """
    return list1 + list2


def concat(lists):
    """
    Recursively concatenate a series of lists into one flattened list.

    Args:
    - lists: A list of lists to be concatenated.

    Returns:
    - A new list containing all elements from the input lists, in order.
    """
    if not lists:
        return []
    return lists[0] + concat(lists[1:])


def filter(function, lst):
    """
    Filter a list based on a predicate function.

    Args:
    - function: A function that returns True for items to keep.
    - lst: The input list to filter.

    Returns:
    - A new list containing only the items for which the predicate returns True.
    """
    return [item for item in lst if function(item)]


def length(lst):
    """
    Recursively calculate the number of items in a list.

    Args:
    - lst: The list to count.

    Returns:
    - int: The number of elements in the list.
    """
    if not lst:
        return 0
    return 1 + length(lst[1:])


def map(func, lst):
    """
    Recursively apply a function to each element of a list and return the list of results.

    Args:
    - func (callable): A function that takes a single element of the list and returns a result.
                       The function should be pure, meaning no side effects, as this implementation
                       assumes only transformations are applied to elements.
    - lst (list): The list to map over. The list can be empty or contain any type of elements
                  compatible with the function.

    Returns:
    - list: A new list of results after applying `func` to each element in `lst`. If `lst` is empty, an empty list is returned.
    """
    if not lst:
        return []
    return [func(lst[0])] + map(func, lst[1:])


def foldl(function, list, accumulator):
    """
    Recursively reduces a list from left to right using a given function and an initial accumulator.

    Args:
    - function: A function that takes the current accumulator and an item, and returns the new accumulator value.
    - list: The list to be reduced.
    - accumulator: The initial value for the accumulator.

    Returns:
    - The final accumulator value after applying the function to each item from left to right.
    """
    if not list:
        return accumulator
    else:
        return foldl(function, list[1:], function(accumulator, list[0]))

def foldr(function, list, accumulator):
    """
    Reduces a list from right to left using a given function and an initial accumulator.

    Args:
    - function: A function that takes the current accumulator and an item, and returns the new accumulator value.
    - list: The list to be reduced.
    - accumulator: The initial value for the accumulator.

    Returns:
    - The final accumulator value after applying the function to each item from right to left.
    """
    for item in reversed(list):
            accumulator = function(accumulator, item)
    return accumulator


def reverse(lst):
    """
    Recursively reverse the order of elements in a list.

    Args:
    - The list to reverse.

    Returns:
    - A new list with the elements of lst in reversed order.
    """
    if not lst:
        return []
    return reverse(lst[1:]) + [lst[0]]
